// @paisabazaar/guc/service/visitService.js
"use strict";

var utility = require('../utility/utility');

function VisitService() {
  this.createVisit = function (config, cookieData, utmData) {
    var vmsUrl = config.vmsUrl,
        siteId = config.siteId;
    var url = vmsUrl;
    var urlParams = utility.urlParams();
    var visitId = utility.uuid();
    var visit = cookieData.visit,
        absoluteParentVisitId = cookieData.absoluteParentVisitId,
        id = cookieData.id;
    var payload = {
      visitId: visitId,
      deviceId: id,
      utmContent: utmData && utmData.utmContent ? utmData.utmContent : utility.result(urlParams, 'utm_content', null),
      utmMedium: utmData && utmData.utmMedium ? utmData.utmMedium : utility.result(urlParams, 'utm_medium', null),
      utmSource: utmData && utmData.utmSource ? utmData.utmSource : utility.result(urlParams, 'utm_source', utility.result(urlParams, 'data_source', null)),
      utmName: utmData && utmData.utmName ? utmData.utmName : utility.result(urlParams, 'utm_name', null),
      utmTerm: utmData && utmData.utmTerm ? utmData.utmTerm : utility.result(urlParams, 'utm_term', null),
      utmTitle: utmData && utmData.utmTitle ? utmData.utmTitle : utility.result(urlParams, 'utm_title', null),
      utm: utmData && utmData.utm ? utmData.utm : utility.result(urlParams, 'utm', null),
      dataSource: utmData && utmData.dataSource ? utmData.dataSource : utility.result(urlParams, 'data_source', null),
      completeUrl: window.location.href,
      createdAt: utility.currentTime(),
      absoluteParentVisitId: absoluteParentVisitId,
      pbUtmCampaign: utmData && utmData.pbUtmCampaign ? utmData.pbUtmCampaign : utility.result(urlParams, 'pbUtmCampaign', null),
      pbUtmContent: utmData && utmData.pbUtmContent ? utmData.pbUtmContent : utility.result(urlParams, 'pbUtmContent', null),
      pbUtmMedium: utmData && utmData.pbUtmMedium ? utmData.pbUtmMedium : utility.result(urlParams, 'pbUtmMedium', null),
      pbUtmSource: utmData && utmData.pbUtmSource ? utmData.pbUtmSource : utility.result(urlParams, 'pbUtmSource', null),
      pbUtmTerm: utmData && utmData.pbUtmSource ? utmData.pbUtmSource : utility.result(urlParams, 'pbUtmSource', null)
    };

    if (visit && visit.visitId) {
      payload.parentVisitId = visit.visitId;
    }

    utility.xhrRequest('POST', url, {}, payload).then(function (result) {})["catch"](function (error) {});
    window.pbgucGtag({
      'label': 'pbguc_create_visit',
      'userType': 'new',
      'site-id': siteId,
      'visitId': visitId,
      'parentVisitId': visit && visit.visitId ? visit.visitId : 'null',
      'absoluteParentVisitId': absoluteParentVisitId
    });
    return {
      visitId: payload.visitId,
      utmSource: payload.utmSource,
      createdAt: payload.createdAt
    };
  };
}

;
module.exports = new VisitService();
