// @paisabazaar/sso-react/webpack.config.js
let path                = require('path');
let webpack             = require('webpack');
let BUILD_DIR           = path.resolve(__dirname);
let APP_DIR             = path.resolve(__dirname);
var PACKAGE             = require('./package.json');

let config = {
    entry           : APP_DIR + '/index.js',

    output: {
        path        : BUILD_DIR,
        filename    : 'sso-plugin.js'
    },

    module : {
        loaders : [
            {
                test    : /\.jsx?/,
                include : APP_DIR,
                loader  : 'babel-loader',
                
            },
            {
                test    : /\.css$/,
                use     : [
                    'style-loader', 
                    'css-loader'
                ]
            },
            {
                test: /\.scss$/,
                use: [
                    "style-loader",
                    "css-loader",
                    "sass-loader"
                ]
            },
            { test: /\.(png|woff|woff2|eot|ttf|svg|jpg|webp)$/, loader: 'url-loader?limit=100000' }

        ]
    },
    plugins: [
        new webpack.NamedModulesPlugin(),
        new webpack.optimize.UglifyJsPlugin({
            compress: { warnings: false },
            comments    : false,
            mangle      : false,
            minimize    : true
        })
    ]
};

module.exports = config;
