// @paisabazaar/guc/utility/GA.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var KafkaService = require('../service/kafkaService');

function loadGTagScript() {
  var existingScript = document.getElementById('pbgucGtagScript');

  if (!existingScript) {
    var headScrEle = document.createElement("script");
    headScrEle.setAttribute("id", 'pbgucGtagScript');
    headScrEle.innerHTML = "(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':\n      new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],\n      j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=\n      'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);\n      })(window,document,'script','dataLayer','GTM-JGPT');";
    document.head.appendChild(headScrEle);
  }

  var existingTag = document.getElementById('gtagIframe');

  if (!existingTag) {
    var scriptEle = document.createElement("noscript");
    scriptEle.setAttribute("id", 'gtagIframe');
    scriptEle.innerHTML = "<iframe src=\"https://www.googletagmanager.com/ns.html?id=GTM-JGPT\" height=\"0\" width=\"0\" style=\"display:none;visibility:hidden\"></iframe>";
    document.body.prepend(scriptEle);
  }

  function gtag(arg) {
    var standardParams = _defineProperty({
      event: 'pbguc',
      category: 'pbguc',
      action: 'PbgucDataUpdated',
      pageType: 'journey',
      label: 'pbguc',
      clientId: '1',
      userType: 'revisit',
      leadType: 'parent',
      lob: "central"
    }, "leadType", "parent");

    window.dataLayer.push(_objectSpread(_objectSpread({}, standardParams), arg));

    try {
      KafkaService.sendToKafka(this.configuration, _objectSpread(_objectSpread({}, standardParams), arg));
    } catch (error) {}
  }

  window.pbgucGtag = gtag;
}

module.exports = loadGTagScript;
